ecdsa/src/signer/rfc6979.rs

```diff
@@ -0,0 +1,121 @@
+//! Support for computing deterministic ECDSA ephemeral scalar (`k`) using
+//! the method described in RFC 6979 rules (Section 3.2):
+//! <https://tools.ietf.org/html/rfc6979#section-3>
+
+#![allow(non_snake_case)]
+
+use elliptic_curve::{
+    generic_array::GenericArray,
+    ops::Invert,
+    scalar::NonZeroScalar,
+    zeroize::{Zeroize, Zeroizing},
+    Arithmetic, ElementBytes, FromBytes,
+};
+use hmac::{Hmac, Mac, NewMac};
+use signature::digest::{BlockInput, Digest, FixedOutput, Reset, Update};
+
+/// Generate ephemeral scalar `k` from the secret scalar and a digest of the
+/// input message.
+pub(super) fn generate_k<C, D>(
+    secret_scalar: &NonZeroScalar<C>,
+    msg_digest: D,
+) -> Zeroizing<NonZeroScalar<C>>
+where
+    C: Arithmetic,
+    C::Scalar: Invert<Output = C::Scalar> + Zeroize,
+    D: BlockInput<BlockSize = C::ElementSize>
+        + FixedOutput<OutputSize = C::ElementSize>
+        + Clone
+        + Default
+        + Reset
+        + Update,
+    ElementBytes<C>: Zeroize,
+{
+    let x = Zeroizing::new(secret_scalar.to_bytes());
+
+    // TODO(tarcieri): don't panic (i.e. unwrap)! add/use trait for reducing digests mod q
+    let h1: ElementBytes<C> = C::Scalar::from_bytes(&msg_digest.finalize())
+        .unwrap()
+        .into();
+
+    let mut hmac_drbg = HmacDrbg::<D>::new(&*x, &h1);
+
+    loop {
+        let k = NonZeroScalar::from_bytes(&hmac_drbg.next());
+
+        if k.is_some().into() {
+            return Zeroizing::new(k.unwrap());
+        }
+    }
+}
+
+/// Internal implementation of `HMAC_DRBG` as described in NIST SP800-90A:
+/// <https://csrc.nist.gov/publications/detail/sp/800-90a/rev-1/final>
+///
+/// This is a HMAC-based deterministic random bit generator used internally
+/// to compute a deterministic ECDSA ephemeral scalar `k`.
+// TODO(tarcieri): generalize and extract this into the `hmac` crate?
+struct HmacDrbg<D>
```

Comment:
@newpavlov it might be interesting to extract or reimplement this in the `hmac` crate (possibly as a `CryptoRng`?)

Reply:
I think it will be better to implement it in a separate crate and it would fit nicely into the currently empty [CSRNGs repository](https://github.com/RustCrypto/CSRNGs). And it looks like @sorpaas has already [published](https://crates.io/crates/hmac-drbg) such crate (although without implementing the `rand_core` traits).

@sorpaas
Would you be interested in transferring `hmac-drbg` crate to this organization?

Reply:
Oh neat! Unfortunately it's `hmac` dependency is a version behind or otherwise it looks like what I need.

Reply:
Opened a PR to bump the `hmac` dependency here: https://github.com/tarcieri/rust-hmac-drbg/pull/1